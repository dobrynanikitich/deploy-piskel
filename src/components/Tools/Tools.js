import React, { Component } from 'react';
import PenSize from './PenSize/PenSize';
import Tool from './Tool/Tool';
import SwitchColor from './SwitchColor/SwitchColor';
import ShotCuts from './ShotCuts/ShotCuts';
import Select from 'react-select';

import classes from './Tools.module.scss';

class Tools extends Component {
    toolsList = [
      {
        id: 'pen',
        iconName: 'pen-fancy',
        tooltip: 'Pen (P)',
        keyCode: 'KeyP'
      },
      {
        id: 'eraser',
        iconName: 'eraser',
        tooltip: 'Eraser (E)',
        keyCode: 'KeyE'
      },
      {
        id: 'stroke',
        iconName: 'pencil-ruler',
        tooltip: 'Stroke (E)',
        keyCode: 'KeyS'
      },
      {
        id: 'bucket',
        iconName: 'fill-drip',
        tooltip: 'Bucket (B)',
        keyCode: 'KeyB'
      },
      {
        id: 'bucketall',
        iconName: 'palette',
        tooltip: 'Bucket all pixels at the same color (A)',
        keyCode: 'KeyA'
      },
  ];

  options = [
    { value: 32, label: '32' },
    { value: 64, label: '64' },
    { value: 128, label: '128' },
  ];

  setActiveTool = (e) => {
    const { id } = e.currentTarget;
    if (id !== this.props.activeTool) {
      this.props.changeState('activeTool', id)
    }
  }

  changePencilSizeHandler = (e) => {
    const { id } = e.currentTarget
    switch (id) {
      case 'firstSize':
        this.props.changeState('pencilSize', 1);
        this.props.changeState('activePenSize', 'firstSize');
        break;
      case 'secondSize':
        this.props.changeState('pencilSize', 2);
        this.props.changeState('activePenSize', 'secondSize');
        break;
      case 'thirdSize':
        this.props.changeState('pencilSize', 3);
        this.props.changeState('activePenSize', 'thirdSize');
        break;
      case 'forthSize':
        this.props.changeState('pencilSize', 4);
        this.props.changeState('activePenSize', 'forthSize');
        break; 
      default: break;
    }
  }

  changeCanvasSizeHandler = (data) => {
    this.props.changeState('pixelsNumber', data.value);
  }

  handleShortcuts = (action, event) => {
    // console.log('++++++')
    // switch(action) {
    //   case 'p': console.log('++++');
    //   break
    //   default: break
    // }
  }

  render() {
    const { changeState, activeTool, activePenSize, currentColor, pixelsNumber } = this.props;

    return (
      <>
        <section className={classes.ToolsSection}>
          <PenSize
            activeId={activePenSize}
            changePencilSizeHandler={this.changePencilSizeHandler}
          />
          <Select
            onChange={this.changeCanvasSizeHandler}
            options={this.options}
            placeholder={pixelsNumber}
            className={classes.SelectInput}
          />
          <div className={classes.MainTools}>
            {this.toolsList.map(tool => (
              <Tool
                id={tool.id}
                key={tool.id}
                active={tool.id === activeTool}
                icon={tool.iconName}
                setActiveTool={this.setActiveTool}
                tooltip={tool.tooltip}
              />
            ))}
          </div>
          <SwitchColor 
            changeState={changeState}
            currentColor={currentColor}
          />
          <ShotCuts />
        </section>
      </>
    );
  };
};

export default Tools;