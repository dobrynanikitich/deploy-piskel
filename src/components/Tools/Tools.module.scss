.ToolsSection {
    width: 100%;
    max-width: 140px;
    display: flex;
    flex-direction: column;
}

.MainTools {
    width: 100%;
    max-width: 135px;
    min-height: 200px;
    display: flex;
    flex-wrap: wrap;
    justify-content: space-between;
    padding: 5px;
    box-sizing: border-box;
}

.SelectInput {
    outline: none;
    width: 100%;
    max-width: 85px;
    margin: 27px 0;
}
