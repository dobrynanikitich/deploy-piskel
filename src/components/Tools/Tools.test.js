import React from 'react';
import Enzyme, { shallow, render, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { createSerializer } from 'enzyme-to-json';
import sinon from 'sinon';

import Tools from './Tools';

expect.addSnapshotSerializer(createSerializer({ mode: "deep" }));

Enzyme.configure({ adapter: new Adapter() });

it('renders correctly', () => {
    const wrapper = shallow(
      <Tools 
        changeState={() => {}}
        activeTool={0}
        activePenSize={'firstSize'}
        currentColor={'#000000'}
        pixelsNumber={32}
        getChildContext={() => {}}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly again', () => {
    const wrapper = render(
      <Tools 
        changeState={() => {}}
        activeTool={0}
        activePenSize={'firstSize'}
        currentColor={'#000000'}
        pixelsNumber={32}
        getChildContext={() => {}}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

it('change state function should be executed', () => {
    const spy = sinon.spy();
    const wrapper = shallow(
      <Tools 
        changeState={spy}
        activeTool={0}
        activePenSize={'firstSize'}
        currentColor={'#000000'}
        pixelsNumber={32}
        getChildContext={() => {}}
      />
    );

    wrapper.instance().setActiveTool({ currentTarget: { id: 1 } });
    expect(spy.calledOnce).toBe(true);
  });

  
it('change state function should not be executed', () => {
    const spy = sinon.spy();
    const wrapper = shallow(
      <Tools 
        changeState={spy}
        activeTool={0}
        activePenSize={'firstSize'}
        currentColor={'#000000'}
        pixelsNumber={32}
        getChildContext={() => {}}
      />
    );

    wrapper.instance().setActiveTool({ currentTarget: { id: 0 } });
    expect(spy.calledOnce).toBe(false);
  });

  it('change state function should be executed with correct params and right quantity', () => {
    const spy = sinon.spy();
    const wrapper = shallow(
      <Tools 
        changeState={spy}
        activeTool={0}
        activePenSize={'firstSize'}
        currentColor={'#000000'}
        pixelsNumber={32}
        getChildContext={() => {}}
      />
    );

    wrapper.instance().changePencilSizeHandler({ currentTarget: { id: 'firstSize' } });
    expect(spy.calledTwice).toBe(true);
  });

  it('change state function should be executed with correct params and right quantity', () => {
    const spy = sinon.spy();
    const wrapper = shallow(
      <Tools 
        changeState={spy}
        activeTool={0}
        activePenSize={'firstSize'}
        currentColor={'#000000'}
        pixelsNumber={32}
        getChildContext={() => {}}
      />
    );

    wrapper.instance().changePencilSizeHandler({ currentTarget: { id: 'secondSize' } });
    expect(spy.calledTwice).toBe(true);
  });

  it('change state function should be executed with correct params and right quantity', () => {
    const spy = sinon.spy();
    const wrapper = shallow(
      <Tools 
        changeState={spy}
        activeTool={0}
        activePenSize={'firstSize'}
        currentColor={'#000000'}
        pixelsNumber={32}
        getChildContext={() => {}}
      />
    );

    wrapper.instance().changePencilSizeHandler({ currentTarget: { id: 'thirdSize' } });
    expect(spy.calledTwice).toBe(true);
  });

  it('change state function should be executed with correct params and right quantity', () => {
    const spy = sinon.spy();
    const wrapper = shallow(
      <Tools 
        changeState={spy}
        activeTool={0}
        activePenSize={'firstSize'}
        currentColor={'#000000'}
        pixelsNumber={32}
        getChildContext={() => {}}
      />
    );

    wrapper.instance().changePencilSizeHandler({ currentTarget: { id: 'forthSize' } });
    expect(spy.calledTwice).toBe(true);
  });

  
  it('change state function should not be executed', () => {
    const spy = sinon.spy();
    const wrapper = shallow(
      <Tools 
        changeState={spy}
        activeTool={0}
        activePenSize={'firstSize'}
        currentColor={'#000000'}
        pixelsNumber={32}
        getChildContext={() => {}}
      />
    );

    wrapper.instance().changePencilSizeHandler({ currentTarget: { id: 'fifthSize' } });
    expect(spy.calledTwice).toBe(false);
  });

  it('change state function should not be executed', () => {
    const spy = sinon.spy();
    const wrapper = shallow(
      <Tools 
        changeState={spy}
        activeTool={0}
        activePenSize={'firstSize'}
        currentColor={'#000000'}
        pixelsNumber={32}
        getChildContext={() => {}}
      />
    );

    wrapper.instance().changeCanvasSizeHandler({ data: { value: 32 } });
    expect(spy.calledOnce).toBe(true);
  });