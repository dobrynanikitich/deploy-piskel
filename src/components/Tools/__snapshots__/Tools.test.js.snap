// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Fragment>
  <section
    className="ToolsSection"
  >
    <PenSize
      activeId="firstSize"
      changePencilSizeHandler={[Function]}
    />
    <StateManager
      className="SelectInput"
      defaultInputValue=""
      defaultMenuIsOpen={false}
      defaultValue={null}
      onChange={[Function]}
      options={
        Array [
          Object {
            "label": "32",
            "value": 32,
          },
          Object {
            "label": "64",
            "value": 64,
          },
          Object {
            "label": "128",
            "value": 128,
          },
        ]
      }
      placeholder={32}
    />
    <div
      className="MainTools"
    >
      <Tool
        active={false}
        icon="pen-fancy"
        id="pen"
        key="pen"
        setActiveTool={[Function]}
        tooltip="Pen (P)"
      />
      <Tool
        active={false}
        icon="eraser"
        id="eraser"
        key="eraser"
        setActiveTool={[Function]}
        tooltip="Eraser (E)"
      />
      <Tool
        active={false}
        icon="pencil-ruler"
        id="stroke"
        key="stroke"
        setActiveTool={[Function]}
        tooltip="Stroke (E)"
      />
      <Tool
        active={false}
        icon="fill-drip"
        id="bucket"
        key="bucket"
        setActiveTool={[Function]}
        tooltip="Bucket (B)"
      />
      <Tool
        active={false}
        icon="palette"
        id="bucketall"
        key="bucketall"
        setActiveTool={[Function]}
        tooltip="Bucket all pixels at the same color (A)"
      />
    </div>
    <SwitchColor
      changeState={[Function]}
      currentColor="#000000"
    />
    <ShotCuts />
  </section>
</Fragment>
`;

exports[`renders correctly again 1`] = `
<section
  class="ToolsSection"
>
  <div
    class="PenSizeMenu"
  >
    <div
      class="PenSize PenSizeActive"
      data-tip="1x"
      id="firstSize"
    >
      <div
        class="firstSize"
      />
    </div>
    <div
      class="PenSize"
      data-tip="2x"
      id="secondSize"
    >
      <div
        class="secondSize"
      />
    </div>
    <div
      class="PenSize"
      data-tip="3x"
      id="thirdSize"
    >
      <div
        class="thirdSize"
      />
    </div>
    <div
      class="PenSize"
      data-tip="4x"
      id="forthSize"
    >
      <div
        class="forthSize"
      />
    </div>
  </div>
  <div
    class="SelectInput css-2b097c-container"
  >
    <div
      class=" css-yk16xz-control"
    >
      <div
        class=" css-1hwfws3"
      >
        <div
          class=" css-1wa3eu0-placeholder"
        >
          32
        </div>
        <div
          class="css-b8ldur-Input"
        >
          <div
            class=""
            style="display:inline-block"
          >
            <input
              aria-autocomplete="list"
              autocapitalize="none"
              autocomplete="off"
              autocorrect="off"
              id="react-select-2-input"
              spellcheck="false"
              style="box-sizing:content-box;width:1px;label:input;background:0;border:0;font-size:inherit;opacity:1;outline:0;padding:0;color:inherit"
              tabindex="0"
              type="text"
              value=""
            />
            <div
              style="position:absolute;top:0;left:0;visibility:hidden;height:0;overflow:scroll;white-space:pre"
            />
          </div>
        </div>
      </div>
      <div
        class=" css-1hb7zxy-IndicatorsContainer"
      >
        <span
          class=" css-1okebmr-indicatorSeparator"
        />
        <div
          aria-hidden="true"
          class=" css-tlfecz-indicatorContainer"
        >
          <svg
            aria-hidden="true"
            class="css-6q0nyr-Svg"
            focusable="false"
            height="20"
            viewBox="0 0 20 20"
            width="20"
          >
            <path
              d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
  <div
    class="MainTools"
  >
    <div
      class="Tool"
      data-tip="Pen (P)"
      id="pen"
    />
    <div
      class="Tool"
      data-tip="Eraser (E)"
      id="eraser"
    />
    <div
      class="Tool"
      data-tip="Stroke (E)"
      id="stroke"
    />
    <div
      class="Tool"
      data-tip="Bucket (B)"
      id="bucket"
    />
    <div
      class="Tool"
      data-tip="Bucket all pixels at the same color (A)"
      id="bucketall"
    />
  </div>
  <div
    class="SwitchColorBlock"
  >
    <label
      class="CurrentColor"
      id="firstColor"
    >
      <input
        id="mainColor"
        type="color"
        value="#808080"
      />
    </label>
    <label
      class="PrevColor"
      id="secondColor"
    >
      <input
        id="secondColor"
        type="color"
        value="#808080"
      />
    </label>
  </div>
  <div
    class="SwitchColorsBtn"
    data-tip="Press to switch colors"
  />
  <div
    class="ShotCutsBlock"
  />
</section>
`;
